### ISSUE TITLE
[bug]: lncli listchaintxns

### ISSUE THREAD

**Issue by electropavuk** [lightningnetwork/lnd#7316]
### Background

From lncli documentation:

```
NAME:
lncli listchaintxns - List transactions from the wallet.

DESCRIPTION:

List all transactions an address of the wallet was involved in.

This call will return a list of wallet related transactions that paid
to an address our wallet controls, or spent utxos that we held. The
start_height and end_height flags can be used to specify an inclusive
block range over which to query for transactions. If the end_height is
less than the start_height, transactions will be queried in reverse.
```

_If the end_height is less than the start_height, transactions will be queried in reverse._

Despite the order of `start_height` and `end_height` the transactions are returned in the same order.

### Your environment

* version of `lnd` `v0.15.1-beta`
* which operating system (`uname -a` on *Nix) `Linux f466fc441522 6.1.4-zen2-1-zen #1 ZEN SMP PREEMPT_DYNAMIC Sat, 07 Jan 2023 15:27:06 +0000 x86_64 Linux`
* version of `bitcoind` `Bitcoin Core RPC client version v22.0.0`

### Steps to reproduce

* start lnd
* mine blocks
* send transaction
* mine blocks
* send transaction
* mine blocks
* call `lncli listchaintxns`

### Expected behaviour

`lncli listchaintxns --start_height 1 --end_height 2000` returns transactions in direct order.
`lncli listchaintxns --start_height 2000 --end_height 1` returns transactions in reversed order.

### Actual behaviour

Both return transactions in reversed order:
```
lncli listchaintxns --start_height 1130 --end_height 1140 | grep block_height
    "block_height": 1140,
    "block_height": 1134,
```

```
lncli listchaintxns --start_height 1140 --end_height 1130 | grep block_height
    "block_height": 1140,
    "block_height": 1134,
```

**Comment by electropavuk** [comment]
I guess the most intuitive behavior would be:

* `start_height <= end_height`, error otherwise
* transactions are returned in direct order (oldest first)
* an additional flag `--reverse` is added to change the order (`--reverse true`, recent first)

**Comment by Chinwendu20** [comment]
Hello please can i work on this?

**Comment by positiveblue** [comment]
@Chinwendu20 ofc, let us know if you have any questions during development + remember to link this issue in the PR

**Comment by Chinwendu20** [comment]
Thanks @positiveblue . I would work on this

**Comment by lndcontrib** [comment]
> I would work on this

It's ok. I'll work on another issue.

**Comment by Chinwendu20** [comment]
Hello all, I might be a bit slow with this but rest assured I am working on it. I am using this as a motivation to get comfortable with the codebase as I aim to apply as a summer of bitcoin intern under this project. I am currently reading documentation to help me become more acquainted with this unfamiliar territory. As soon as I am comfortable I would submit a PR or ask for help but at least from a place of clarity then.

**Comment by lucasdcf** [comment]
Take your time @Chinwendu20

Feel free to contact me on Telegram if you need help: @lucasdcf

**Comment by Chinwendu20** [comment]
Thanks @lucasdcf . I think the reported "bug" is intended. For reference see the function called in the listchaintxn function

https://github.com/lightningnetwork/lnd/blob/983d917d4a05e1c2a45f5c89c4cf2951ffa22746/lnrpc/rpc_utils.go#L88-L99
Sorting is actually done by number of confirmations
I think we just need to modify the documentation

**Comment by Chinwendu20** [comment]
Hello I think this 'bug' was intended, see this snippet of code from a function being called by the listchaintxns.

https://github.com/lightningnetwork/lnd/blob/983d917d4a05e1c2a45f5c89c4cf2951ffa22746/lnrpc/rpc_utils.go#L88-L99

I think we just need to modify the docs.

**Comment by 9rib** [comment]
> Thanks @lucasdcf . Hello I think this 'bug' was intended, see this snippet of code from a function being called by the listchaintxns.
> 
> https://github.com/lightningnetwork/lnd/blob/983d917d4a05e1c2a45f5c89c4cf2951ffa22746/lnrpc/rpc_utils.go#L88-L99
> 
> I think we just need to modify the docs.

Good point. If this is the case, modifying the docs seems like a reasonable idea.

**Comment by positiveblue** [comment]
I may be missing something here but even if we are sorting by number of confirmations rather than height `lncli listchaintxns --start_height 1130 --end_height 1140 | grep block_height`  should be different than `lncli listchaintxns --start_height 1140 --end_height 1130 | grep block_height` when we have two entries with ` "block_height": 1140` and `"block_height": 1134`

**Comment by 9rib** [comment]
> I may be missing something here but even if we are sorting by number of confirmations rather than height `lncli listchaintxns --start_height 1130 --end_height 1140 | grep block_height` should be different than `lncli listchaintxns --start_height 1140 --end_height 1130 | grep block_height` when we have two entries with ` "block_height": 1140` and `"block_height": 1134`

Why the outputs should be different? A transaction with `"block_height": 1134` will always have more confirmations than a transaction with `"block_height": 1140`. So the later comes first anyway.

**Comment by positiveblue** [comment]
yes, but then there will never be any difference between `--start_height {low_high} --end_height {high_height} ` and ` `--start_height {high_height} --end_height {low_height} `

I think it would be better to sort by `height`  and always put the `non-confirmed` first for example.

We can leave the API consumers to do the last step and sort it however they want (while we always return them sorted by confirmation order) but then we should return an error if `--start_height > --end_height`

**Comment by 9rib** [comment]
I think that's ok, we should return an error in the second case.
> yes, but then there will never be any difference between `--start_height {low_high} --end_height {high_height} ` and `--start_height {high_height} --end_height {low_height}`

I think this would be unclear: non-confirmed transactions first, and right after that very old transactions.
> I think it would be better to sort by `height` and always put the `non-confirmed` first for example.

I guess this is the best solution.
> We can leave the API consumers to do the last step and sort it however they want (while we always return them sorted by confirmation order) but then we should return an error if `--start_height > --end_height` + fix the docs.

Do you mean sorted by the number of confirmations, as it is done now? + error + docs
> while we always return them sorted by confirmation order

**Comment by Chinwendu20** [comment]
I still think there is a use case for users to see confirmed transactions first before non-confirmed transactions i.e. start_height > end_height. I do not think we should return error in that case. We can just remove the bit where there is the same output in both cases. I guess that was the intention of this issue, just that when I saw something contrary boldly commented in the codebase, I thought that might be the behavior agreed upon. I am going through the commit history to see if there is a reason it was kept that way

**Comment by 9rib** [comment]
In this case, we should add an additional flag, this will be much clearer than controlling order by swapping `start_height` and  `end_height`.
> I still think there is a use case for users to see confirmed transactions first before non-confirmed transactions i.e.

As were proposed earlier.
> I guess the most intuitive behavior would be:
> 
> * `start_height <= end_height`, error otherwise
> * transactions are returned in direct order (oldest first)
> * an additional flag `--reverse` is added to change the order (`--reverse true`, recent first)

But considering implementation details, I think it would be sufficient and more clear to add an error and change the docs.
> // Sort transactions by number of confirmations rather than height

**Comment by Chinwendu20** [comment]
Oh okay I will implement it as discussed

**Comment by mohamedawnallah** [comment]
#7496

**Comment by kornpow** [comment]
I might try and take this on since it seems like the PR has been abandoned.

**Comment by kornpow** [comment]
After taking a look at this code related to this bug, and also rebasing @Chinwendu20's work onto master and trying it out, I am at a crossroads. The api has been updated since this last discussion, and so some of the ideas aren't as relevant anymore.

1. `start_height` and `end_height` is now order enforced. There is no reversal functionality implied anymore in the docs. if `end_height` < `start_height` lncli sends an error message.
2. Since rebasing this code, the `--reverse` flag works, but I am not sure if we want to do this anymore. This api function has a lot more control knobs now, `index_offset` and `max_transactions`.
3. It seems to me that it might be better to just leave sorting as a task for the consumer, and close this issue.

What do you think @guggero ?

**Comment by guggero** [comment]
The final version of https://github.com/lightningnetwork/lnd/pull/8998 still fetched all transactions from the wallet, then just used the `index_offset` and `max_transactions` values to slice into the result.
So if we add a `--reverse` flag, we'd just reverse the slice before the offset/limit operations are applied.
Start and end height would just limit the initial query, so `--reverse` would only operate on the result after applying the height limits.

**Comment by saraogiraj94** [comment]
Hi @guggero — based on the discussion here and PR #9558 which added input validation blocking start_height > end_height, it looks like the path forward is adding an explicit --reverse flag to listchaintxns rather than relying on height ordering to imply reversal.

My understanding of the correct implementation:

Add a reverse bool field to GetTransactionsRequest in the proto
Pass it through ListTransactionDetails in lnwallet/btcwallet/btcwallet.go
Reverse the transaction slice before pagination (index_offset / max_transactions) is applied
Is this the right approach? If so, I'd like to take this up.

**Comment by Vandit1604** [comment]
I want to take a dig at this. No open PR right now — #10579 closed for inactivity, not on the design, and #8998/#9558 never landed the feature.

@saraogiraj94, you outlined the fix on 2026-05-01 (reverse flag on the request, reverse before `index_offset`/`max_transactions` pagination) but haven't opened a PR. Still on it? If not, I'll run with your approach.

@guggero, is an explicit `--reverse` flag the direction you want, over inferring reversal from height order? #9558 already rejects `start_height > end_height`, so a flag looks like the clean way to expose it. Want to confirm before I write code.

**Comment by Vandit1604** [comment]
Reproduced on master (a912dd698): `listchaintxns --start_height 401 --end_height 403` returns newest-first (403, 402, 401), and swapping the heights to reverse is now rejected by #9558. No way to get oldest-first.

The guard is CLI-only — `rpcServer.GetTransactions` passes heights straight to `ListTransactionDetails`, no validation. So `--reverse` should be a `GetTransactionsRequest` field, reversing the slice before pagination, to fix the RPC too.
